File: framework/languages/dotnet/e2e/ResilienceLifecycle/Client/Scenarios/RlC3NodePauseRecoveryScenario.cs
// Verifies RL-C3 normal process stop and topology recovery.
using ResilienceLifecycle.Client.Support;
using ResilienceLifecycle.Shared;
using Zlink.HttpClient;

namespace ResilienceLifecycle.Client.Scenarios;

// RL-C3 verifies normal SIGTERM restart with a new owner generation.
internal static class RlC3NodePauseRecoveryScenario
{
    public static async Task RunAsync(
        ZLinkHttpClient consumer,
        ZLinkHttpClient registry,
        ResilienceProcessManager processes,
        ZLinkHttpClient providerA,
        ZLinkHttpClient providerB)
    {
        var oldRows = (await registry.Post("/topology/wait")
            .Body(new TopologyWaitReq("api-b", "Ready", 1))
            .Async<TopologyEntryRes[]>()).Body;
        var oldGeneration = oldRows.Single().Generation;

        await processes.StopProviderBWithSigtermAsync();
        await registry.Post("/topology/wait")
            .Body(new TopologyWaitReq("api-b", "Ready", 0))
            .Async<TopologyEntryRes[]>();
        await ProviderTrafficProbe.WaitUntilProviderExcludedAsync(
            consumer,
            "api-b",
            "rl-c3-converge",
            "RL-C3");

        var during = (await consumer.Post("/profile/request")
            .Body(new ProfileReq("fast", "rl-c3-during-down"))
            .Async<ProfileRes>()).Body;
        ZlinkStreamAssert.Ensure(during.ProviderRid == "api-a", "RL-C3 did not use surviving provider during node down.");

        var connectionCount = (await consumer.Get("/connections").Async<string[]>()).Body.Length;
        var restarted = await processes.StartProviderBAsync();
        var recoveredRows = (await registry.Post("/topology/wait")
            .Body(new TopologyWaitReq("api-b", "Ready", 1))
            .Async<TopologyEntryRes[]>()).Body;
        ZlinkStreamAssert.Ensure(
            recoveredRows.Length == 1 && recoveredRows[0].Generation != oldGeneration,
            "RL-C3 did not converge to exactly one new owner generation.");
        await consumer.Post("/connections/wait")
            .Body(new ConnectionWaitReq(
                ["kind=ConnectionReady", $"remote={restarted.Endpoint}"], connectionCount))
            .Async<string[]>();
        await ProviderTrafficProbe.DriveUntilProviderServesAsync(
            consumer,
            providerB,
            "rl-c3-recovered",
            "RL-C3",
            "profile-request|rid=api-b|marker=rl-c3-recovered-");

        {
            using var timeout = new CancellationTokenSource(TimeSpan.FromSeconds(15));
            var waitA = providerA.Post("/evidence/wait").Body(new EvidenceWaitReq(["marker=rl-c3-during-down"], []))
                .Async<string[]>(timeout.Token).AsTask();
            var waitB = providerB.Post("/evidence/wait").Body(new EvidenceWaitReq(["marker=rl-c3-during-down"], []))
                .Async<string[]>(timeout.Token).AsTask();
            var completed = await Task.WhenAny(waitA, waitB);
            var evidence = (await completed).Body;
            timeout.Cancel();
            ZlinkStreamAssert.Ensure(
                evidence.Any(line => line.Contains("marker=rl-c3-during-down", StringComparison.Ordinal)),
                "RL-C3 did not record expected evidence 'marker=rl-c3-during-down'.");
        }
        {
            using var timeout = new CancellationTokenSource(TimeSpan.FromSeconds(15));
            var waitA = providerA.Post("/evidence/wait")
                .Body(new EvidenceWaitReq(["profile-request|rid=api-b|marker=rl-c3-recovered-"], []))
                .Async<string[]>(timeout.Token).AsTask();
            var waitB = providerB.Post("/evidence/wait")
                .Body(new EvidenceWaitReq(["profile-request|rid=api-b|marker=rl-c3-recovered-"], []))
                .Async<string[]>(timeout.Token).AsTask();
            var completed = await Task.WhenAny(waitA, waitB);
            var evidence = (await completed).Body;
            timeout.Cancel();
            ZlinkStreamAssert.Ensure(
                evidence.Any(line =>
                    line.Contains("profile-request|rid=api-b|marker=rl-c3-recovered-", StringComparison.Ordinal)),
                "RL-C3 did not record expected evidence 'marker=rl-c3-recovered-'.");
        }

        Console.WriteLine("scenario RL-C3 passed");
    }

}

File: framework/languages/node/e2e/ResilienceLifecycle/Client/Scenarios/rl-c3-node-pause-recovery-scenario.ts
// RL-C3: 정상 restart 뒤 새 lifecycle로 수렴한다 시나리오를 검증한다.
import type { ProfileRes } from '../../Shared/messages';
import type { ClientOptions } from '../Support/client-options';
import { postJson } from '../../../http-client';
import { startProvider } from '../Support/managed-provider';
import {
  profileReq,
  waitForProviderTraffic,
  waitPeerPresent,
  waitUntilAvailable,
  waitUntilDown
} from '../Support/resilience-helpers';
import { ensure } from '../Support/scenario-assert';
import type { ScenarioState } from '../Support/scenario-state';

export async function runRlC3(options: ClientOptions, state: ScenarioState): Promise<void> {
  if (state.providerBProcess !== undefined) {
    await state.providerBProcess.stop();
    state.providerBProcess = undefined;
  } else {
    await postJson(options.providerBUrl, '/shutdown');
  }
  await waitUntilDown(options.providerBUrl);

  const during = await postJson<ProfileRes>(
    options.consumerUrl,
    '/profile/request',
    profileReq('rl-c3-during-down')
  );
  ensure(during.providerRid === 'api-a', 'RL-C3 did not use surviving provider during node down.');

  const restarted = startProvider({
    providerMain: options.providerMain,
    logDir: options.logDir,
    redisEndpoint: options.redisEndpoint,
    redisKeyPrefix: options.redisKeyPrefix,
    name: 'api-b-c3-restored',
    rid: 'api-b',
    httpUrl: options.providerBUrl,
    channelEndpoint: options.providerBChannelEndpoint,
    evidenceFileName: 'api-b-c3-restored.evidence.log'
  });
  await restarted.waitReady();
  state.providerBProcess = restarted;
  await waitUntilAvailable(options.providerBUrl);
  await waitPeerPresent(options.peerLocationUrl, 'api-b');

  await waitForProviderTraffic(options.consumerUrl, 'rl-c3-recovered', 'api-b');

  const downEvidence = await postJson<string[]>(options.providerAUrl, '/evidence/wait', {
    contains: 'marker=rl-c3-during-down'
  });
  ensure(
    downEvidence.some((line) => line.includes('marker=rl-c3-during-down')),
    'RL-C3 did not record expected during-down evidence.'
  );

  const recoveredEvidence = await postJson<string[]>(options.providerBUrl, '/evidence/wait', {
    contains: 'profile-request|rid=api-b|value=fast|marker=rl-c3-recovered-'
  });
  ensure(
    recoveredEvidence.some((line) => line.includes('profile-request|rid=api-b') && line.includes('marker=rl-c3-recovered-')),
    'RL-C3 did not record expected recovered api-b evidence.'
  );

  console.log('scenario RL-C3 passed');
}
